benches/bin/mvcctxn.rs

```diff
@@ -0,0 +1,476 @@
+    for key in keys {
+        txn.commit(key, commit_ts).unwrap();
+    }
+    do_write(&*db, txn.into_modifies());
+}
+
+fn prepare_test_db(versions: usize, value_len: usize, keys: &[Vec<u8>], path: &str) -> Arc<DB> {
+    let mut config = DbConfig::default();
+    // Use a huge write_buffer_size to avoid flushing data to disk.
+    config.defaultcf.write_buffer_size = ReadableSize::gb(1);
+    config.writecf.write_buffer_size = ReadableSize::gb(1);
+    config.lockcf.write_buffer_size = ReadableSize::gb(1);
+
+    let cf_ops = vec![
+        CFOptions::new(CF_DEFAULT, config.defaultcf.build_opt()),
+        CFOptions::new(CF_LOCK, config.lockcf.build_opt()),
+        CFOptions::new(CF_WRITE, config.writecf.build_opt()),
+        CFOptions::new(CF_RAFT, config.raftcf.build_opt()),
+    ];
+
+    let db = Arc::new(new_engine_opt(path, DBOptions::new(), cf_ops).unwrap());
+
+    for _ in 0..versions {
+        for key in keys {
+            let value = vec![0u8; value_len];
+            let start_ts = next_ts();
+            let commit_ts = next_ts();
+
+            prewrite(
+                db.clone(),
+                &[Mutation::Put((Key::from_raw(key), value))],
+                key,
+                start_ts,
+            );
+            commit(db.clone(), &[Key::from_raw(key)], start_ts, commit_ts);
+        }
+    }
+    db
+}
+
+#[inline]
+fn get(db: Arc<DB>, key: &Key, statistics: &mut Statistics) -> Option<Vec<u8>> {
+    let snapshot = get_snapshot(db.clone());
+    let start_ts = next_ts();
+    let snapstore = SnapshotStore::new(snapshot, start_ts, IsolationLevel::SI, true);
+    snapstore.get(key, statistics).unwrap()
+}
+
+fn bench_get(db: Arc<DB>, keys: &[Vec<u8>]) -> f64 {
+    let mut fake_statistics = Statistics::default();
+    let mut rng = rand::thread_rng();
+    do_bench(
+        || {
+            let index = rng.gen_range(0, keys.len());
+            let key = Key::from_raw(&keys[index]);
+
+            get(db.clone(), &key, &mut fake_statistics).unwrap()
+        },
+        500000,
+    )
+}
+
+fn bench_set(db: Arc<DB>, keys: &[Vec<u8>], value_len: usize) -> f64 {
+    let mut rng = rand::thread_rng();
+    do_bench(
+        || {
+            let start_ts = next_ts();
+            let commit_ts = next_ts();
+            let value = vec![0u8; value_len];
+
+            let key = &keys[rng.gen_range(0, keys.len())];
+
+            prewrite(
+                db.clone(),
+                &[Mutation::Put((Key::from_raw(key), value))],
+                key,
+                start_ts,
+            );
+            commit(db.clone(), &[Key::from_raw(key)], start_ts, commit_ts)
+        },
+        500000,
+    )
+}
+
+fn bench_delete(db: Arc<DB>, keys: &[Vec<u8>]) -> f64 {
+    let mut rng = rand::thread_rng();
+    do_bench(
+        || {
+            let start_ts = next_ts();
+            let commit_ts = next_ts();
+
+            let key = &keys[rng.gen_range(0, keys.len())];
```

Comment:
Some rows may have been already deleted?

Reply:
I didn't find any logic  in `MvccTxn` that handles deleting a already-deleted key, but simply insert a delete record in `CF_WRITE`. So I think it's ok to do this.